content/ai/concepts/search-algorithms/terms/breadth-first-search/breadth-first-search.md

```diff
@@ -0,0 +1,116 @@
+---
+Title: 'Breadth-First Search'
+Description: 'A shortest-path and traversing algorithm for unweighted graphs.'
+Subjects:
+  - 'Code Foundations'
+  - 'Computer Science'
+Tags:
+  - 'Graphs'
+  - 'Graph Search'
+  - 'Algorithms'
+CatalogContent:
+  - 'learn-python'
+  - 'paths/computer-science'
+---
+
+**Breadth-first search** is a shortest-path and traversing algorithm for unweighted graphs. This is a foundational algorithm in [graph](https://www.codecademy.com/resources/docs/general/graph) theory from which many other algorithms start.
+
+## Features
+
+Some of the features and constraints that define the use and functionality of a breadth-first search (BFS) algorithm include the following:
+
+- A relatively simple and efficient algorithm for finding the shortest-path in graphs that do not have edge weights.
+- In addition to finding the shortest path between two nodes this algorithm can be used to map or traverse a graph to find all connections or if a given path exists between any pair of nodes.
+
+## Implementation
+
+The example below walks through a basic implementation of BFS that will take a starting node, an end node, and a graph as arguments. However, to start, a brief overview of the necessary steps or pseudocode will serve as an outline for the final implementation.
+
+The BFS algorithm is built upon a simple process of elimination. It begins with the selection of a node and the identification of all the neighboring nodes. The algorithm iterates through these basic steps until all nodes have been evaluated (or another objective has been achieved). Through this process, the graph is explored in a layered approach where each node at a given level (or degree) is evaluated before moving on to the next. Implicitly, a search employing this process will yield the shortest path between any two nodes.
+
+As noted in the description above, the BFS process is an iterative one. The implementation commonly revolves around adding and removing nodes as their evaluated from a stack.
+
+The implementation below can be broken down into the following steps:
```

Comment:
A few things to note about this procedure:

- I think we should mention that the graph here is represented as an edge list as opposed to other ways of representing a graph (adj-list, adj-matrix).
- I think we should just say "using a queue" as opposed to "using a list in a queue-like manner". It would be easier to understand that way.
- It is unclear to me how the algorithm determines which nodes are the starting node and destination node.
- I also don't understand how to use a variable to track a path. I've always used a map or a dictionary to map a node to the node that put it in the queue for processing so that when the destination node is found, we can use the map to backtrack and find the path. 
- The last few steps seem specific to the example being used. If a destination node isn't found, the algorithm can either return an empty list, -1, a string, or raise an error indicating failure depending on what the programmer needs.
- I also don't see a mention of how we keep track of nodes that have already been added to the queue to prevent them from being added again.

Reply:
- I would say that the implementation of edgeview (the edges attribute) isn't quite an edge list. I see this as more of a hybrid implementation.
- I didn't want to say using a queue because I'm not actually using a queue.
- The starting and ending nodes are set as arguments, and the first iteration of the function compares all the edge pairs to the starting node. 
- I agree that what you describe, backtracking to assemble the path, is perhaps the more common procedure. However, I don't see why we should necessarily remain faithful to that version. I would recommend adding a print statement, print(curr_path) , after the append to queue statement to visualize how the function steps through the bfs cycle and tracks the path concurrently .
- The outputs are a choice, I don't think there is anything wrong with the implementation as it is. If someone wants to use it they can modify the returns to suit their needs.
- The list 'tested_nodes' is used to track nodes previously visited.

Reply:
Ahh, I see. But `tested_nodes` is only visible in the code. I feel like a non-Python programmer may just be interested in the algorithm and not an implementation in Python so they may skip the code.

Reply:
I've added this step to the pseudo code